email-service/internal/handlers/rate.go

```diff
@@ -0,0 +1,72 @@
+package handlers
+
+import (
+	"context"
+	"log/slog"
+	"time"
+
+	"github.com/GenesisEducationKyiv/software-engineering-school-4-0-Hukyl/email-service/internal/models"
+)
+
+type rateRepository interface {
+	Create(rate *models.Rate) error
+}
+
+func doWithContext(ctx context.Context, f func() error) error {
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+		if err := f(); err != nil {
+			getLogger().Error("error", slog.Any("error", err))
+		}
+		getLogger().Debug("function done", slog.Any("function", f))
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			err := ctx.Err()
+			getLogger().Error("context done", slog.Any("error", err))
+			return err
+		case <-done:
+			getLogger().Debug("done")
+			return nil
+		}
+	}
+}
+
+type RateEvents struct {
+	repo rateRepository
+}
+
+func (s *RateEvents) SaveRate(
+	ctx context.Context,
+	from, to string, value float32, time time.Time,
+) error {
+	getLogger().Info(
```

Comment:
Coming back to my previous comments with log's context.

You can do it this way:


```go
log := getLogger().With(
		slog.Any("from", from),
		slog.Any("to", to),
		slog.Any("rate", value),
		slog.Any("time", time),
)

log.Info("rate fetched")
```

and in next calls you can use this `log` variable in the log function, and it will automatically log `from` `to` `rate` and `time` fields. This can help you in debugging in the future.